//
// This file is part of InfoGrid(tm). You may not use this file except in
// compliance with the InfoGrid license. The InfoGrid license and important
// disclaimers are contained in the file LICENSE.InfoGrid.txt that you should
// have received with InfoGrid. If you have not received LICENSE.InfoGrid.txt
// or you do not consent to all aspects of the license and the disclaimers,
// no license is granted; do not use this file.
//
// For more information about InfoGrid go to http://infogrid.org/
//
// Copyright 1998-2010 by R-Objects Inc. dba NetMesh Inc., Johannes Ernst
// All rights reserved.
//

package org.infogrid.probe.test;

import java.io.IOException;
import java.net.URISyntaxException;
import java.text.ParseException;
import org.infogrid.mesh.MeshObject;
import org.infogrid.mesh.MeshObjectIdentifierNotUniqueException;
import org.infogrid.mesh.NotPermittedException;
import org.infogrid.mesh.RelatedAlreadyException;
import org.infogrid.meshbase.net.CoherenceSpecification;
import org.infogrid.meshbase.net.NetMeshBaseIdentifier;
import org.infogrid.meshbase.net.NetMeshObjectIdentifierFactory;
import org.infogrid.meshbase.transaction.TransactionException;
import org.infogrid.module.ModuleException;
import org.infogrid.probe.ApiProbe;
import org.infogrid.probe.ProbeException;
import org.infogrid.probe.StagingMeshBase;
import org.infogrid.probe.StagingMeshBaseLifecycleManager;
import org.infogrid.util.instrument.InstrumentedThread;
import org.infogrid.util.logging.Log;

/**
 * Tests concurrent access to the same Probe.
 */
public class ProbeTest9a
    extends
        AbstractProbeTest9
{
    /**
     * Run the test.
     *
     * @throws Exception all sorts of things may happen during a test
     */
    public void run()
        throws
            Exception
    {
        probeInvocationCounter = 0;

        //

        
        log.info( "Creating threads" );

        Runnable [] runnables = new Runnable[] {
                new MyRunnable( test1_URL ) {
                    protected void recordResult(
                            MeshObject result )
                    {
                        foundObject1 = result;
                    }
                    protected void recordException(
                            Throwable t )
                    {
                        throwable1 = t;
                    }
                },
                new MyRunnable( test1_URL ) {
                    protected void recordResult(
                            MeshObject result )
                    {
                        foundObject2 = result;
                    }
                    protected void recordException(
                            Throwable t )
                    {
                        throwable2 = t;
                    }
                },
        };

        InstrumentedThread [] threads = createThreads( runnables, new String[] {
                    "one",
                    "two"
                } );

        checkNotObject( foundObject1, "found object 1 too early" );
        checkNotObject( foundObject2, "found object 2 too early" );
        checkNotObject( throwable1,   "unexpected exception 1" );
        checkNotObject( throwable2,   "unexpected exception 2" );

        sleepFor( 100L );
        
        //

        log.info( "Advancing first thread" );

        threads[0].advanceTo( location1 );
        checkNotObject( foundObject1, "found object 1 too early" );
        checkNotObject( foundObject2, "found object 2 too early" );
        checkNotObject( throwable1,   "unexpected exception 1" );
        checkNotObject( throwable2,   "unexpected exception 2" );

        sleepFor( PINGPONG_ROUNDTRIP_DURATION );

        log.info( "Advancing second thread" );

        boolean successful = threads[1].advanceTo( location1, 2000L );
        checkCondition( !successful, "thread2 should not have been able to advance to breakpoint" );
        checkNotObject( foundObject1, "found object 1 too early" );
        checkNotObject( foundObject2, "found object 2 too early" );
        checkNotObject( throwable1,   "unexpected exception 1" );
        checkNotObject( throwable2,   "unexpected exception 2" );

        //

        log.info( "second try for thread2" );

        successful = threads[1].advanceTo( location1, 2000L );
        checkCondition( !successful, "thread2 should not have been able to advance to breakpoint" );

        checkNotObject( foundObject1, "found object 1 too early" );
        checkNotObject( foundObject2, "found object 2 too early" );
        checkNotObject( throwable1,   "unexpected exception 1" );
        checkNotObject( throwable2,   "unexpected exception 2" );

        //
        
        log.info( "Completing threads" );

        threads[0].completeThreadAndWait();
        threads[1].completeThreadAndWait();

        Thread.sleep( 2000L ); // give it a little so threads can finish

        checkObject( foundObject1, "did not find object 1" );
        checkObject( foundObject2, "did not find object 2" );
        checkIdentity( foundObject1, foundObject2, "Not the same instance" );
        checkNotObject( throwable1,   "unexpected exception 1" );
        checkNotObject( throwable2,   "unexpected exception 2" );

        checkEquals( probeInvocationCounter, 1, "invoked wrong number of times" );

        //

        log.info( "Make sure we only have 1 Shadow" );

        checkEquals( theMeshBase.getShadowMeshBases().size(),         1, "wrong number of shadows" );
        checkEquals( countFromIterator( theMeshBase.proxies(), log ), 1, "wrong number of proxies" );
        checkEquals( countFromIterator( theMeshBase.getShadowMeshBaseFor( test1_URL ).proxies(), log ), 1, "wrong number of proxies" );
    }

    /**
     * Main program.
     *
     * @param args command-line arguments
     */
    public static void main(
            String [] args )
    {
        ProbeTest9a test = null;
        try {
            if( args.length > 0 ) {
                System.err.println( "Synopsis: <no args>" );
                System.err.println( "aborting ..." );
                System.exit( 1 );
            }

            test = new ProbeTest9a( args );
            test.run();

        } catch( Throwable ex ) {
            log.error( ex );
            System.exit(1);
        }
        if( test != null ) {
            test.cleanup();
        }
        if( errorCount == 0 ) {
            log.info( "PASS" );
        } else {
            log.info( "FAIL (" + errorCount + " errors)" );
        }
        System.exit( errorCount );
    }

    /**
     * Constructor.
     *
     * @param args command-line arguments
     * @throws Exception all sorts of things may happen during a test
     */
    public ProbeTest9a(
            String [] args )
        throws
            Exception
    {
        super( ProbeTest9a.class, TestApiProbeA.class );
    }

    // Our Logger
    private static Log log = Log.getLogInstance( ProbeTest9a.class );

    /**
     * The test Probe.
     */
    public static class TestApiProbeA
        implements
            ApiProbe
    {
        public void readFromApi(
                NetMeshBaseIdentifier  networkId,
                CoherenceSpecification coherence,
                StagingMeshBase        mb )
            throws
                MeshObjectIdentifierNotUniqueException,
                RelatedAlreadyException,
                TransactionException,
                NotPermittedException,
                ProbeException,
                IOException,
                ModuleException,
                URISyntaxException,
                ParseException
        {
            log.debug( getClass().getName() + ".readFromApi()" );

            ++probeInvocationCounter;

            MeshObject                      home   = mb.getHomeObject();
            StagingMeshBaseLifecycleManager life   = mb.getMeshBaseLifecycleManager();
            NetMeshObjectIdentifierFactory  idFact = mb.getMeshObjectIdentifierFactory();

            MeshObject one = life.createMeshObject( idFact.fromExternalForm( "one" ));
            MeshObject two = life.createMeshObject( idFact.fromExternalForm( "two" ));

            home.relate( one );

            try {
                location1.reached();

            } catch( InterruptedException ex ) {
                throw new ProbeException.Other( networkId, ex );
            }

            home.relate( two );
        }
    }
}
